lib/model/recent_senders.dart

```diff
@@ -0,0 +1,133 @@
+import 'package:collection/collection.dart';
+import 'package:flutter/foundation.dart';
+
+import '../api/model/model.dart';
+
+class MessageIdTracker {
+  MessageIdTracker();
+
+  @visibleForTesting
+  MessageIdTracker.fromIds(List<int> ids) {
+    _ids.addAll([...{...ids}]..sort());
+  }
+
+  final List<int> _ids = [];
+
+  void add(int id) {
+    if (_ids.isEmpty) {
+      _ids.add(id);
+    } else {
+      int i = lowerBound(_ids, id);
+      if (i < _ids.length && _ids[i] == id) return; // the [id] already exists, so do not add it.
```

Comment:
I think that refactor would provide a more descriptive tone to the code, but I am not sure if it is necessary for this stage as we would end up using it just in one place (in `add()` method). We have a similar method `RecentDmConversationsView._insertSorted`, but that's used in two places. Maybe @gnprice would like to provide their suggestion here.

Reply:
I agree it'd be good to make that information more explicit. I think a good main way to do that would be by writing down how the data structure `_ids` is supposed to work — the fact that it's meant to be sorted, and have only distinct elements. Like so:
```dart
  /// A list of distinct message IDs, sorted.
  final List<int> _ids = [];
```

Once the reader knows that, it's mostly clear already what `add` should do; the only remaining nuance is that it's fine if the message ID is already there, and it just won't do anything in that case. So:
```dart
  /// Add the given message ID to the list, if not already present.
  void add(int id) {
```

In general, the meanings of the data structures are often the highest-impact facts to write down about how an area of code is intended to work. There's an often-quoted line about this [by Fred Brooks](https://en.wikiquote.org/wiki/Fred_Brooks), one of the earliest software-engineering managers:
> Show me your flowcharts and conceal your tables, and I shall continue to be mystified. Show me your tables, and I won’t usually need your flowcharts; they’ll be obvious.

The reference to "flowcharts" shows how old the observation is (published 1975), but it remains accurate today.

Reply:
BTW I'd encourage asking questions directed at me in #mobile-team, if it's a GitHub thread I'm not already active in. (Could even ask here first, and then just quote it in a message there.) I try to periodically catch up on all threads in the repo, but I skim and might miss things (almost missed this one), and I'm also likely to see it much sooner in Zulip.

Reply:
> Show me your flowcharts and conceal your tables, and I shall continue to be mystified. Show me your tables, and I won’t usually need your flowcharts; they’ll be obvious
Thanks for sharing this. I had never came across this before!!

What do you think of making the variable name more descriptive instead of using comment? Comments tend to get outdated.
```
final List<int> _distinctSortedMessageIDs = [];
```

Reply:
> making the variable name more descriptive instead of using comment? Comments tend to get outdated.

I definitely sympathize with that direction in general. But this doesn't feel to me like the right situation for it.

I think a key part of what's giving me that feeling is that this is the central data structure that all the code surrounding it is structured around. If those aspects (the values are distinct, they're sorted) were to change in some future commit, it'd look more like scrapping and replacing this whole class than like making small tweaks to it. So that makes the info that goes into a piece of docs on this field very stable.